mod-clans/src/main/java/gg/solarmc/loader/clans/Clan.java

```diff
@@ -0,0 +1,449 @@
+        } else {
+            manager.insertAllianceCache(this.getID(),receiver.getID());
+            return true;
+        }
+    }
+
+    /**
+     * Revokes ally. Can be called from either clan. Revokes alliance for both.
+     * @param transaction the tx
+     * @return true if ok, false if one of the clans or both of the clans has no ally
+     * @throws IllegalArgumentException if the clan does not have an ally
+     */
+    public boolean revokeAlly(Transaction transaction) {
+        Optional<Clan> localClan = this.getAlliedClan(transaction);
+
+        if (localClan.isEmpty()) throw new IllegalArgumentException("This clan has no ally!");
+
+        int i = transaction.getProperty(DSLContext.class)
+                .deleteFrom(CLANS_CLAN_ALLIANCES)
+                .where(CLANS_CLAN_ALLIANCES.CLAN_ID.eq(this.clanID))
+                .or(CLANS_CLAN_ALLIANCES.ALLY_ID.eq(this.clanID))
+                .execute();
+
+        if (i != 2) {
+            return false;
+        } else {
+            manager.invalidateAllianceCache(this.getID(),localClan.orElseThrow().getID());
+            return true;
+        }
+    }
+
+    /**
+     * Marks clan as an enemy
+     * @param transaction twix bar
+     * @param receiver the clan to mark as an enemy.
+     * @return true if the enemy was added, false if the clans are already enemies
+     * @throws IllegalArgumentException if the presented clan is an ally or is the same object (what?)
+     */
+    public boolean addClanAsEnemy(Transaction transaction, Clan receiver) {
+        if (receiver.getID() == this.clanID) throw new IllegalArgumentException("Tried to mark clan enemy as same clan!");
+
+        this.getAlliedClan(transaction).ifPresent(ally -> {
+            if (receiver.getID() == ally.getID()) throw new IllegalArgumentException("Tried to add allied clan as enemy");
+        });
+
+        int res = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_ENEMIES)
+                .columns(CLANS_CLAN_ENEMIES.CLAN_ID,CLANS_CLAN_ENEMIES.ENEMY_ID)
+                .values(this.clanID,receiver.getID())
+                .onDuplicateKeyIgnore()
+                .execute();
+
+        return res == 1;
+    }
+
+    /**
+     * Marks clan enemy as no longer an enemy
+     * @param transaction the tx
+     * @param clan clan provided
+     * @throws IllegalArgumentException if passed self or if passed ally
+     * @return true if went ok, false if method executing clan is not enemies with provided clan
+     */
+    public boolean removeClanAsEnemy(Transaction transaction, Clan clan) {
+        if (clan.getID() == this.clanID) throw new IllegalArgumentException("Tried to unmark self as enemy??");
+
+        this.getAlliedClan(transaction).ifPresent(ally -> {
+            if (clan.getID() == ally.getID()) throw new IllegalArgumentException("Tried to add allied clan as enemy");
+        });
+
+        ClansClanEnemiesRecord rec = transaction.getProperty(DSLContext.class)
```

Comment:
Retrieving the ClansClanEnemiesRecord costs 2 database queries. You should perform a DELETE FROM to do this in a single query.